pkg/controller/gc/controller.go

```diff
@@ -0,0 +1,178 @@
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+
+	gcc := &GCController{
+		kubeClient: kubeClient,
+		podControl: controller.RealPodControl{
+			Recorder:   eventBroadcaster.NewRecorder(api.EventSource{Component: "pod-garbage-collector"}),
+			KubeClient: kubeClient,
+		},
+	}
+
+	gcc.threshold = 5
+
+	gcc.podStore.Store, gcc.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return gcc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return gcc.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		fullResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			UpdateFunc: gcc.updatePod,
+			DeleteFunc: gcc.deletePod,
+		},
+	)
+	return gcc
+}
+
+func (gcc *GCController) Run(stop <-chan struct{}) {
+	go gcc.podController.Run(stop)
+	go util.Until(gcc.work, time.Second, stop)
+	<-stop
+	gcc.queue.ShutDown()
+}
+
+func (gcc *GCController) work() {
+	for {
+		_, quit := gcc.queue.Get()
+		if quit {
+			return
+		}
+		gcc.gc()
+	}
+}
+
+func (gcc *GCController) gc() {
+	markedPods := make([]*api.Pod, 0)
+	expectation, _, _ := gcc.expectations.GetExpectations(gcControllerKey)
+	_, expectedDeleteCount := expectation.GetExpectations()
+	pods, _ := gcc.podStore.List(labels.Everything())
+	activeCount := len(pods)
+	for _, pod := range pods {
+		if isPodTerminal(pod) {
+			markedPods = append(markedPods, pod)
+		}
+	}
+	sort.Sort(byCreationTimestamp(markedPods))
+
+	// is the type assertion safe?
+	deleteCount := activeCount - gcc.threshold - int(expectedDeleteCount)
+
+	if deleteCount > len(markedPods) {
+		deleteCount = len(markedPods)
+	}
+
+	if deleteCount > 0 {
+		glog.Infof("garbage collecting %v pods", deleteCount)
+	}
+
+	var wait sync.WaitGroup
+	gcc.expectations.ExpectDeletions(gcControllerKey, deleteCount)
+	for i := 0; i < deleteCount; i++ {
+		wait.Add(1)
+		go func(toDelete *api.Pod) {
+			defer wait.Done()
+			if err := gcc.podControl.DeletePod(toDelete.Namespace, toDelete.Name); err != nil {
+				defer util.HandleError(err)
+				gcc.expectations.DeletionObserved(gcControllerKey)
+			}
+		}(markedPods[i])
+	}
+	wait.Wait()
+}
+
+func (gcc *GCController) updatePod(oldObj, newObj interface{}) {
+	oldPod := oldObj.(*api.Pod)
+	newPod := newObj.(*api.Pod)
+	// if pod becomes terminal, trigger GC run
+	if !isPodTerminal(oldPod) && isPodTerminal(newPod) {
+		gcc.queue.Add(struct{}{})
+	}
+}
+
+func (gcc *GCController) deletePod(obj interface{}) {
+	pod := obj.(*api.Pod)
+	if isPodTerminal(pod) {
+		gcc.expectations.DeletionObserved(gcControllerKey)
+	}
+}
+
+func isPodTerminal(pod *api.Pod) bool {
+	phase := pod.Status.Phase
+	return phase == api.PodSucceeded || phase == api.PodFailed
```

Comment:
Other future terminal conditions - pod deletion timestamp >> system time (should be covered by node controller), pod active deadline seconds >> pod start time + deadline, pod is on node that no longer exists (admin deleted node, node controller did not observe deletion).

Also, this condition needs a delta from some system time, because a job framework may own the pods deletion after a terminal state and needs to observe completion.  I think as is this breaks job controller.